File: scripts.old/bot/deploy.py
## brownie run scripts/deploy.py --network n9997


from brownie import * 
from brownie.convert.datatypes import EthAddress 
import os
import json
import yaml
import time

from dotenv import load_dotenv ##pipx install python-dotenv 

load_dotenv()

def brownie_project(tag):
    with open("brownie-config.yaml", "r") as brownie_config:
        config_dict = yaml.load(brownie_config, Loader=yaml.FullLoader)
        for entry in config_dict["dependencies"]:
            if tag in entry:
                return  project.load(entry)
    return project.main

def updatemapjson( network, title, address):
    open("./build/deployments/map.json", "a+")
    with open("./build/deployments/map.json", "r+") as map_json:
        map_json.seek(0)
        try:
            data = json.load(map_json)
        except:
            data = json.loads("{}")
        list = []
        if data.get(network) is None:
            data[network] = dict()
        if not data[network].get(title) is None:
            list = data[network][title]
            if list[len(list)-1] == address:
                return
        ##if list[-1]!=address:
        list.append(address)
        data[network][title] = list
        map_json.seek(0)
        map_json.write(json.dumps(data,sort_keys=True,indent=2))


def find_network(chainid):
    with open("network-config.json", "r") as network_config:
        data = json.load(network_config)
        for entry in data["live"]:
            for network in entry["networks"]:
                print(network)
                if network["chainid"] == chainid:
                    return network["host"]

def export_network():
    print(f"\nExporting network")
    with open("network-config.yaml", "r") as brownie_config:
        config_dict = yaml.load(brownie_config, Loader=yaml.FullLoader)
        with open("network-config.json", "w") as brownie_config_json:
            json.dump(config_dict, brownie_config_json,sort_keys=True,indent=2)

def supply(accountfrom,account,amount):
    if account.balance() < amount:
        accountfrom.transfer(account,amount-account.balance()).wait(1)


EthAddress0 = EthAddress("0x0000000000000000000000000000000000000000")

def mainaccount():
    if network.show_active() in {"development", "ganache"}:
        return accounts[0]
    else:
        return accounts.add(os.getenv("MAIN_PKEY"))

def extraaccount():
    if network.show_active() in {"development", "ganache"}:
        return accounts[1]
    else:
        return accounts.add(os.getenv("EXTRA_PKEY"))

def supply(accountfrom,account,amount):
    if account.balance() < amount:
        accountfrom.transfer(account,amount-account.balance()).wait(1)
    assert(account.balance() >= amount)


LocatorAddress = EthAddress("0x455b153B592d4411dCf5129643123639dcF3c806")
mycontracts = brownie_project("bffr05/contracts")
token = brownie_project("bffr05/token")
kAccess = network.web3.keccak(bytes("Access", 'ascii'))
kTokenPlace = network.web3.keccak(bytes("TokenPlace", 'ascii'))
kOracle = network.web3.keccak(bytes("Oracle", 'ascii'))

kForwarder_Deployer = network.web3.keccak(bytes("Forwarder_Deployer", 'ascii'))
kMEquivFT = network.web3.keccak(bytes("MEquivFT", 'ascii'))
kMEquivNFT = network.web3.keccak(bytes("MEquivNFT", 'ascii'))
kGate = network.web3.keccak(bytes("Gate", 'ascii'))
kMulticall = network.web3.keccak(bytes("Multicall", 'ascii'))

myLocator = EthAddress0

def getNoLocator():
    return (len(network.web3.eth.get_code(LocatorAddress)) == 0)
def getLocator():
    nolocator = getNoLocator()
    if nolocator and len(mycontracts.Locator) > 0:
        return mycontracts.Locator[-1].address
    elif nolocator:
        mycontracts.Locator.deploy({"from": mainaccount()})
        return mycontracts.Locator[-1].address
    else:
        return LocatorAddress
    

def main():
    export_network()
    nolocator = getNoLocator()
    myLocator = getLocator()

    print(f"LocatorAddress = {myLocator}")
    print(f"nolocator = {nolocator}")
    print(f"len = {len(network.web3.eth.get_code(myLocator))}")

    updatemapjson(str(network.web3.chain_id),"Locator",myLocator)
    print()

    ILocator = interface.ILocator(myLocator)

    TestAccess = False



    if True: ##len(Access) ==0 or ILocator.get(kAccess) != Access[-1]:
        addr = mycontracts.Access.deploy({"from": mainaccount()})
        ILocator.set['bytes32,address'](kAccess,addr,{"from": mainaccount()})

        if nolocator:
            addr.setLocator(myLocator,{"from": mainaccount()}) 
            assert(addr.locator()==myLocator)
        else:
            assert(addr.locator()==LocatorAddress)

        if TestAccess:
            addr.setOperator(LocatorAddress,True,{"from": mainaccount()})
            assert(addr.isOperator(mainaccount(),LocatorAddress))
    updatemapjson(str(network.web3.chain_id),"Access",ILocator.get(kAccess))


    if True: ##len(TokenPlace) ==0 or ILocator.get(kTokenPlace) != TokenPlace[-1]:
        addr = token.TokenPlace.deploy({"from": mainaccount()})
        ILocator.set['bytes32,address'](kTokenPlace,addr,{"from": mainaccount()})
        ##isTrusted
        if nolocator:
            addr.setLocator(myLocator,{"from": mainaccount()}) 
            assert(addr.locator()==myLocator)
        else:
            assert(addr.locator()==LocatorAddress)

        assert(addr.referral()==ILocator.location(kAccess))

        if False: ##TestAccess:
            assert(not addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,True,{"from": mainaccount()})
            assert(addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,False,{"from": mainaccount()})
            assert(not addr.isOperator(mainaccount(),LocatorAddress))

        mycontracts.interface.IRTrustable(ILocator.location(kAccess)).setContract(addr,True,{"from": mainaccount()}) 

        if TestAccess:
            assert(addr.isOperator(mainaccount(),LocatorAddress))
    updatemapjson(str(network.web3.chain_id),"TokenPlace",ILocator.get(kTokenPlace))


    if True: ##len(TokenPlace) ==0 or ILocator.get(kTokenPlace) != TokenPlace[-1]:
        addr = mycontracts.Oracle.deploy({"from": mainaccount()})
        ILocator.set['bytes32,address'](kOracle,addr,{"from": mainaccount()})
        ##isTrusted
        if nolocator:
            addr.setLocator(myLocator,{"from": mainaccount()}) 
            assert(addr.locator()==myLocator)
        else:
            assert(addr.locator()==LocatorAddress)

        assert(addr.referral()==ILocator.location(kAccess))

        if False: ##TestAccess:
            assert(not addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,True,{"from": mainaccount()})
            assert(addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,False,{"from": mainaccount()})
            assert(not addr.isOperator(mainaccount(),LocatorAddress))

        mycontracts.interface.IRTrustable(ILocator.location(kAccess)).setContract(addr,True,{"from": mainaccount()}) 

        if TestAccess:
            assert(addr.isOperator(mainaccount(),LocatorAddress))
    updatemapjson(str(network.web3.chain_id),"Oracle",ILocator.get(kOracle))

    if True: ##len(Forwarder_Deployer) ==0 or ILocator.get(kForwarder_Deployer) != Forwarder_Deployer[-1]:
        addr = mycontracts.Forwarder_Deployer.deploy({"from": mainaccount()})
        ILocator.set['bytes32,address'](kForwarder_Deployer,addr,{"from": mainaccount()})
    updatemapjson(str(network.web3.chain_id),"Forwarder_Deployer",ILocator.get(kForwarder_Deployer))



    if True: ##len(MEquivFT) ==0 or ILocator.get(kMEquivFT) != MEquivFT[-1]:
        addr = token.MEquivFT.deploy({"from": mainaccount()})
        ILocator.set['bytes32,address'](kMEquivFT,addr,{"from": mainaccount()})

        if nolocator:
            addr.setLocator(myLocator,{"from": mainaccount()})
            assert(addr.locator()==myLocator)
        else:
            assert(addr.locator()==LocatorAddress)

        if False: ##TestAccess:
            assert(not addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,True,{"from": mainaccount()})
            assert(addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,False,{"from": mainaccount()})
            assert(not addr.isOperator(mainaccount(),LocatorAddress))

        mycontracts.interface.IRTrustable(ILocator.location(kAccess)).setContract(addr,True,{"from": mainaccount()}) 

        if TestAccess:
            assert(addr.isOperator(mainaccount(),LocatorAddress))
    updatemapjson(str(network.web3.chain_id),"MEquivFT",ILocator.get(kMEquivFT))


    if True: ##len(MEquivNFT) ==0 or ILocator.get(kMEquivNFT) != MEquivNFT[-1]:
        addr = token.MEquivNFT.deploy({"from": mainaccount()})
        ILocator.set['bytes32,address'](kMEquivNFT,addr,{"from": mainaccount()})

        if nolocator:
            addr.setLocator(myLocator,{"from": mainaccount()})
            assert(addr.locator()==myLocator)
        else:
            assert(addr.locator()==LocatorAddress)

        if False: ##TestAccess:
            assert(not addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,True,{"from": mainaccount()})
            assert(addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,False,{"from": mainaccount()})
            assert(not addr.isOperator(mainaccount(),LocatorAddress))

        mycontracts.interface.IRTrustable(ILocator.location(kAccess)).setContract(addr,True,{"from": mainaccount()}) 

        if TestAccess:
            assert(addr.isOperator(mainaccount(),LocatorAddress))
    updatemapjson(str(network.web3.chain_id),"MEquivNFT",ILocator.get(kMEquivNFT))


    if True: ##len(Gate) ==0 or ILocator.get(kGate) != Gate[-1]:
        GateLib.deploy({"from": mainaccount()})
        addr = Gate.deploy({"from": mainaccount()})
        ILocator.set['bytes32,address'](kGate,addr,{"from": mainaccount()})

        if nolocator:
            addr.setLocator(myLocator,{"from": mainaccount()})
            assert(addr.locator()==myLocator)
        else:
            assert(addr.locator()==LocatorAddress)

        if False: ##TestAccess:
            assert(not addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,True,{"from": mainaccount()})
            assert(addr.isOperator(mainaccount(),LocatorAddress))
            addr.setOperator(LocatorAddress,False,{"from": mainaccount()})
            assert(not addr.isOperator(mainaccount(),LocatorAddress))

        mycontracts.interface.IRTrustable(ILocator.location(kAccess)).setContract(addr,True,{"from": mainaccount()}) 

        if TestAccess:
            assert(addr.isOperator(mainaccount(),LocatorAddress))
    updatemapjson(str(network.web3.chain_id),"Gate",ILocator.get(kGate))

    if False: ##len(Gate) ==0 or ILocator.get(kGate) != Gate[-1]:
        addr = multicall.deploy({"from": mainaccount()})
        ILocator.set(kMulticall,addr,{"from": mainaccount()})
        updatemapjson(str(network.web3.chain_id),"Multicall",ILocator.get['bytes32'](kMulticall))

    ##destChainId = 9998
    ##if web3.eth.chain_id == 9998:
    ##    destChainId = 9997

    ##Gate[-1].tokenOut(destChainId,"0x0000000000000000000000000000000000000000",0,3,{"from": mainaccount(),"value":8})

    time.sleep(0.5)

File: scripts/deploy.js

const hre = require("hardhat");
const assert = require("assert");

require("dotenv").config(); // yarn add dotenv
require('hardhat-deploy'); // yarn add hardhat-deploy npm install -D hardhat-deploy

let LocatorAddress = "0x455b153B592d4411dCf5129643123639dcF3c806";
const EthAddress0 = "0x0000000000000000000000000000000000000000";

const kAccess = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("Access"));
const kTokenPlace = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("TokenPlace"));
const kOracle = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("Oracle"));
const kForwarder_Deployer = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("Forwarder_Deployer"));
const kMEquivFT = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("MEquivFT"));
const kMEquivNFT = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("MEquivNFT"));
const kGate = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("Gate"));
const kMulticall = hre.ethers.utils.keccak256(hre.ethers.utils.toUtf8Bytes("Multicall"));

myLocator = EthAddress0

async function makebalance(from,to,balance)
{
    totransfer = balance - ethers.utils.formatEther(await to.getBalance());
    // console.log(`totransfer = ${ethers.utils.formatEther(ethers.utils.parseEther(totransfer.toString()))}`)
    if (totransfer <=0) 
        return;
        tresp = await from.sendTransaction(
            {
                to: to.address,
                value: ethers.utils.parseEther("1.0")
            })
        await tresp.wait(1)
    // console.log(`balance = ${ethers.utils.formatEther(await to.getBalance())}`)

}



async function main() {
  console.log(`kAccess = ${kAccess}`);
  console.log(`kTokenPlace = ${kTokenPlace}`);
  console.log(`kOracle = ${kOracle}`);
  console.log(`kForwarder_Deployer = ${kForwarder_Deployer}`);
  console.log(`kMEquivFT = ${kMEquivFT}`);
  console.log(`kMEquivNFT = ${kMEquivNFT}`);
  console.log(`kGate = ${kGate}`);
  console.log(`kMulticall = ${kMulticall}`);
  console.log("\n")


  const signer = await ethers.getSigner();
  console.log(`process.env.MAIN_PKEY = ${process.env.MAIN_PKEY}`);
  const mainuser = new ethers.Wallet(process.env.MAIN_PKEY,ethers.provider)
  console.log("\n")


  console.log(`signer=${signer.address} balance=${ethers.utils.formatEther(await signer.getBalance())} nonce=${await signer.getTransactionCount()}`)
  await makebalance(signer,mainuser,1.0)
  console.log(`mainuser=${mainuser.address} balance=${ethers.utils.formatEther(await mainuser.getBalance())} nonce=${await mainuser.getTransactionCount()}`)
  console.log("\n")




  let nolocator = (await ethers.provider.getCode(LocatorAddress)).length<=2;
  if (nolocator) {
    const Locator = await ethers.getContractFactory("Locator",mainuser);
    const locator = await Locator.deploy();
    console.log(`deploying Locator`);
    await locator.deployed();
    console.log("Locator deployed");
    LocatorAddress = locator.address;
  }
  const locator = await ethers.getContractAt("ILocator", LocatorAddress,mainuser);
  console.log(`LocatorAddress=${LocatorAddress}`);
  console.log("\n");


  if (true) {
    const Access = await ethers.getContractFactory("Access",mainuser);
    const access = await Access.deploy();
    console.log(`deploying Access`);
    await access.deployed();
    console.log("Access deployed");

    if (nolocator) {
      await access.setLocator(LocatorAddress);
      assert((await access.locator())==LocatorAddress);
    }
    console.log(`Access Locator=${await access.locator()}`);

    await locator["set(bytes32,address)"](kAccess,access.address);
  }
  const access = await ethers.getContractAt("IRTrustable", await locator.get(kAccess),mainuser);
  console.log(`Locator(Access)=${await locator.get(kAccess)}`);
  console.log("\n");


  if (true) {
    const TokenPlace = await ethers.getContractFactory("TokenPlace",mainuser);
    const tokenplace = await TokenPlace.deploy();
    console.log(`deploying TokenPlace`);
    await tokenplace.deployed();
    console.log("TokenPlace deployed");

    if (nolocator) {
      await tokenplace.setLocator(LocatorAddress);
      assert((await tokenplace.locator())==LocatorAddress);
    }
    console.log(`TokenPlace Locator=${await tokenplace.locator()}`);
    console.log(`TokenPlace referral=${await tokenplace.referral()}`);
    assert((await tokenplace.referral())==(await locator.get(kAccess)));

    await locator["set(bytes32,address)"](kTokenPlace,tokenplace.address);
    await access.setContract(tokenplace.address,true) ;

  }
  console.log(`Locator(TokenPlace)=${await locator.get(kTokenPlace)}`);
  console.log("\n");


  if (true) {
    const Oracle = await ethers.getContractFactory("Oracle",mainuser);
    const oracle = await Oracle.deploy();
    console.log(`deploying Oracle`);
    await oracle.deployed();
    console.log("Oracle deployed");

    if (nolocator) {
      await oracle.setLocator(LocatorAddress);
      assert((await oracle.locator())==LocatorAddress);
    }
    console.log(`Oracle Locator=${await oracle.locator()}`);
    console.log(`Oracle referral=${await oracle.referral()}`);
    assert((await oracle.referral())==(await locator.get(kAccess)));

    await locator["set(bytes32,address)"](kOracle,oracle.address);
    await access.setContract(oracle.address,true) ;

  }
  console.log(`Locator(Oracle)=${await locator.get(kOracle)}`);
  console.log("\n");


  if (true) {
    const Forwarder_Deployer = await ethers.getContractFactory("Forwarder_Deployer",mainuser);
    const forwarder = await Forwarder_Deployer.deploy();
    console.log(`deploying Forwarder_Deployer`);
    await forwarder.deployed();
    console.log("Forwarder_Deployer deployed");

    await locator["set(bytes32,address)"](kForwarder_Deployer,forwarder.address);

  }
  console.log(`Locator(Forwarder_Deployer)=${await locator.get(kForwarder_Deployer)}`);
  console.log("\n");



  if (true) {
    const MEquivNFT = await ethers.getContractFactory("MEquivNFT",mainuser);
    const nft = await MEquivNFT.deploy();
    console.log(`deploying MEquivNFT`);
    await nft.deployed();
    console.log("MEquivNFT deployed");

    if (nolocator) {
      await nft.setLocator(LocatorAddress);
      assert((await nft.locator())==LocatorAddress);
    }
    console.log(`MEquivNFT Locator=${await nft.locator()}`);
    console.log(`MEquivNFT referral=${await nft.referral()}`);
    assert((await nft.referral())==(await locator.get(kAccess)));

    await locator["set(bytes32,address)"](kMEquivNFT,nft.address);
    await access.setContract(nft.address,true) ;

  }
  console.log(`Locator(MEquivNFT)=${await locator.get(kMEquivNFT)}`);
  console.log("\n");


  if (true) {
    const MEquivFT = await ethers.getContractFactory("MEquivFT",mainuser);
    const ft = await MEquivFT.deploy();
    console.log(`deploying MEquivFT`);
    await ft.deployed();
    console.log("MEquivFT deployed");

    if (nolocator) {
      await ft.setLocator(LocatorAddress);
      assert((await ft.locator())==LocatorAddress);
    }
    console.log(`MEquivFT Locator=${await ft.locator()}`);
    console.log(`MEquivFT referral=${await ft.referral()}`);
    assert((await ft.referral())==(await locator.get(kAccess)));

    await locator["set(bytes32,address)"](kMEquivFT,ft.address);
    await access.setContract(ft.address,true) ;

  }
  console.log(`Locator(MEquivFT)=${await locator.get(kMEquivFT)}`);
  console.log("\n");


  if (true) {

    const GateLibTransfer = await ethers.getContractFactory("GateLibTransfer",mainuser);
    const gatelibtransfer = await GateLibTransfer.deploy();
    console.log(`deploying GateLibTransfer`);
    await gatelibtransfer.deployed();
    console.log("GateLibTransfer deployed");

    const GateLibTranscript = await ethers.getContractFactory("GateLibTranscript",mainuser);
    const gatelibtranscript = await GateLibTranscript.deploy();
    console.log(`deploying GateLibTranscript`);
    await gatelibtranscript.deployed();
    console.log("GateLibTranscript deployed");
    
    const GateLibUtil = await ethers.getContractFactory("GateLibUtil",{ signer: mainuser, libraries: { GateLibTransfer: gatelibtransfer.address,GateLibTranscript:gatelibtranscript.address } });
    const gatelibutil = await GateLibUtil.deploy();
    console.log(`deploying GateLibUtil`);
    await gatelibutil.deployed();
    console.log("GateLibUtil deployed");

    const GateLibProcessMessage = await ethers.getContractFactory("GateLibProcessMessage",{ signer: mainuser, libraries: { GateLibUtil: gatelibutil.address, GateLibTransfer: gatelibtransfer.address,GateLibTranscript:gatelibtranscript.address } });
    const gatelibprocessmessage = await GateLibProcessMessage.deploy();
    console.log(`deploying GateLibProcessMessage`);
    await gatelibprocessmessage.deployed();
    console.log("GateLibProcessMessage deployed");



    const Gate = await ethers.getContractFactory("Gate",{ signer: mainuser, libraries: { GateLibProcessMessage: gatelibprocessmessage.address, GateLibUtil: gatelibutil.address, GateLibTransfer: gatelibtransfer.address,GateLibTranscript:gatelibtranscript.address } });
    const gate = await Gate.deploy();
    console.log(`deploying Gate`);
    await gate.deployed();
    console.log("Gate deployed");

    if (nolocator) {
      await gate.setLocator(LocatorAddress);
      assert((await gate.locator())==LocatorAddress);
    }
    console.log(`Gate Locator=${await gate.locator()}`);
    console.log(`Gate referral=${await gate.referral()}`);
    assert((await gate.referral())==(await locator.get(kAccess)));

    await locator["set(bytes32,address)"](kGate,gate.address);
    await access.setContract(gate.address,true) ;

  }
  console.log(`Locator(Gate)=${await locator.get(kGate)}`);
  console.log("\n");

}



// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
